bqskit/passes/processing/treescan.py

```diff
@@ -0,0 +1,226 @@
+"""This module implements the ScanningGateRemovalPass."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+from typing import Callable
+
+from bqskit.compiler.basepass import BasePass
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.operation import Operation
+from bqskit.ir.opt.cost.functions import HilbertSchmidtResidualsGenerator
+from bqskit.ir.opt.cost.generator import CostFunctionGenerator
+from bqskit.runtime import get_runtime
+from bqskit.utils.typing import is_integer
+from bqskit.utils.typing import is_real_number
+
+_logger = logging.getLogger(__name__)
+
+
+class TreeScanningGateRemovalPass(BasePass):
+    """
+    The ScanningGateRemovalPass class.
+
+    Starting from one side of the circuit, run the following:
+
+    Split the circuit operations into chunks of size tree_depth
+    At every iteration:
+    a. Look at the next chunk of operations
+    b. Generate 2 ^ tree_depth circuits. Each circuit corresponds to every
+    combination of whether or not to include one of the operations in the chunk.
+    c. Instantiate in parallel all 2^tree_depth circuits
+    d. Choose the circuit that has the least number of operations and move
+    on to the next chunk of operations.
+
+    This optimization is less greedy than the current ScanningGate removal,
+    which we see can offermuch better quality circuits than ScanningGate.
+    In very rare occasions, ScanningGate may be able to outperform
```

Comment:
I don't understand why ScanningGate would be able to outperform TreeScan. Isn't every candidate considered during ScanningGate considered during TreeScan? I know there is always variance in instantiation so that anything can happen for sure, but is there something I am missing?

Reply:
For example, if there are 20 gates and TreeScan looks at 5 at a time. Let's say that the options are that if you remove the first gate, you can also remove the last 10 gates in the circuit. Otherwise, if you don't remove the first gate, you can remove the next 4 gates in the circuit (but this removes your ability to remove the last 10 gates in the circuit). TreeScan intially would pick the 4 gate option over the single gate option, and then miss out removing the last 10 gates. I don't know if this is even possible in practice, and I haven't seen a single case where this does happen, but AFAIK it could be a possibility?

Reply:
I see. Yeah, that makes sense. Thanks for explaining.